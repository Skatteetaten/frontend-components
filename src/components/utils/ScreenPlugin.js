import { useEffect, useState } from 'react';

const getValuesForComparison = sizeObj => ({
  gt: sizeObj.gt,
  lt: sizeObj.lt,
  sm: sizeObj.sm,
  md: sizeObj.md,
  lg: sizeObj.lg,
  xl: sizeObj.xl,
  xxl: sizeObj.xxl,
  xxxl: sizeObj.xxxl
});

export const UseScreen = () => {
  const [size, setSize] = useState({
    width: 0,
    height: 0,
    sizes: {
      sm: 480,
      md: 640,
      lg: 1024,
      xl: 1366,
      xxl: 1920
    },
    lt: {
      sm: false,
      md: true,
      lg: true,
      xl: true,
      xxl: true,
      xxxl: true
    },
    gt: {
      sm: false,
      md: false,
      lg: false,
      xl: false,
      xxl: false,
      xxxl: false
    },
    sm: true,
    md: false,
    lg: false,
    xl: false,
    xxl: false,
    xxxl: false
  });

  useEffect(() => {
    const getSizes = () => {
      const newSizes = {
        gt: {},
        lt: {}
      };
      if (window.innerHeight !== size.height) {
        newSizes.height = window.innerHeight;
      }

      const w = window.innerWidth;

      if (w !== size.width) {
        newSizes.width = w;
      }

      const s = size.sizes;

      newSizes.gt.sm = w >= s.sm;
      newSizes.gt.md = w >= s.md;
      newSizes.gt.lg = w >= s.lg;
      newSizes.gt.xl = w >= s.xl;
      newSizes.gt.xxl = w >= s.xxl;
      newSizes.lt.sm = false;
      newSizes.lt.md = w < s.sm;
      newSizes.lt.lg = w < s.md;
      newSizes.lt.xl = w < s.lg;
      newSizes.lt.xxl = w < s.xl;
      newSizes.lt.xxxl = w < s.xxl;
      newSizes.sm = !newSizes.gt.sm;
      newSizes.md = newSizes.gt.sm && newSizes.lt.lg;
      newSizes.lg = newSizes.gt.md && newSizes.lt.xl;
      newSizes.xl = newSizes.gt.lg && newSizes.lt.xxl;
      newSizes.xxl = newSizes.gt.xl && newSizes.lt.xxxl;
      newSizes.xxxl = newSizes.gt.xxl;

      const sizeHasChanged =
        JSON.stringify(getValuesForComparison(newSizes)) !==
        JSON.stringify(getValuesForComparison(size));

      return [sizeHasChanged, { sizes: size.sizes, ...newSizes }];
    };

    function updateSizes() {
      const [sizeHasChanged, sizes] = getSizes();
      if (sizeHasChanged) {
        setSize(sizes);
      }
    }

    updateSizes();
    window.addEventListener('resize', updateSizes);
    return () => window.removeEventListener('resize', updateSizes);
  }, [size]);

  return size;
};
