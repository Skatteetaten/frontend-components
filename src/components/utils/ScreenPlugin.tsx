import { useEffect, useState } from 'react';

// @ts-ignore TODO
const getValuesForComparison = sizeObj => ({
  gt: sizeObj.gt,
  lt: sizeObj.lt,
  sm: sizeObj.sm,
  md: sizeObj.md,
  lg: sizeObj.lg,
  xl: sizeObj.xl,
  xxl: sizeObj.xxl,
  xxxl: sizeObj.xxxl
});

export const UseScreen = () => {
  const [size, setSize] = useState({
    width: 0,
    height: 0,
    sizes: {
      sm: 480,
      md: 640,
      lg: 1024,
      xl: 1366,
      xxl: 1920
    },
    lt: {
      sm: false,
      md: true,
      lg: true,
      xl: true,
      xxl: true,
      xxxl: true
    },
    gt: {
      sm: false,
      md: false,
      lg: false,
      xl: false,
      xxl: false,
      xxxl: false
    },
    sm: true,
    md: false,
    lg: false,
    xl: false,
    xxl: false,
    xxxl: false
  });

  useEffect(() => {
    const getSizes = () => {
      const newSizes = {
        gt: {},
        lt: {},
        height: 0,
        width: 0
      };
      if (window.innerHeight !== size.height) {
        newSizes.height = window.innerHeight;
      }

      const w = window.innerWidth;

      if (w !== size.width) {
        newSizes.width = w;
      }

      const s = size.sizes;
      // @ts-ignore TODO
      newSizes.gt.sm = w >= s.sm;
      // @ts-ignore TODO
      newSizes.gt.md = w >= s.md;
      // @ts-ignore TODO
      newSizes.gt.lg = w >= s.lg;
      // @ts-ignore TODO
      newSizes.gt.xl = w >= s.xl;
      // @ts-ignore TODO
      newSizes.gt.xxl = w >= s.xxl;
      // @ts-ignore TODO
      newSizes.lt.sm = false;
      // @ts-ignore TODO
      newSizes.lt.md = w < s.sm;
      // @ts-ignore TODO
      newSizes.lt.lg = w < s.md;
      // @ts-ignore TODO
      newSizes.lt.xl = w < s.lg;
      // @ts-ignore TODO
      newSizes.lt.xxl = w < s.xl;
      // @ts-ignore TODO
      newSizes.lt.xxxl = w < s.xxl;
      // @ts-ignore TODO
      newSizes.sm = !newSizes.gt.sm;
      // @ts-ignore TODO
      newSizes.md = newSizes.gt.sm && newSizes.lt.lg;
      // @ts-ignore TODO
      newSizes.lg = newSizes.gt.md && newSizes.lt.xl;
      // @ts-ignore TODO
      newSizes.xl = newSizes.gt.lg && newSizes.lt.xxl;
      // @ts-ignore TODO
      newSizes.xxl = newSizes.gt.xl && newSizes.lt.xxxl;
      // @ts-ignore TODO
      newSizes.xxxl = newSizes.gt.xxl;

      const sizeHasChanged =
        JSON.stringify(getValuesForComparison(newSizes)) !==
        JSON.stringify(getValuesForComparison(size));

      return [sizeHasChanged, { sizes: size.sizes, ...newSizes }];
    };

    function updateSizes() {
      const [sizeHasChanged, sizes] = getSizes();
      if (sizeHasChanged) {
        // @ts-ignore TODO
        setSize(sizes);
      }
    }

    updateSizes();
    window.addEventListener('resize', updateSizes);
    return () => window.removeEventListener('resize', updateSizes);
  }, [size]);

  return size;
};
