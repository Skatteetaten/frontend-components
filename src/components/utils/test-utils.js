import renderer from 'react-test-renderer';

export const matches = (children) =>
  expect(renderer.create(children).toJSON()).toMatchSnapshot();
