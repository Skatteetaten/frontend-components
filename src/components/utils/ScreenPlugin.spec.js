import React from 'react';
import { UseScreen } from './ScreenPlugin';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';

describe('Test av ScreenPlugin', () => {
  const TestHook = () => {
    const size = UseScreen();
    return (
      <div>
        {size.md && <span>Vises når skjermstørrelsen er 'md'</span>}
        {size.lt.md && (
          <span>Vises når skjermstørrelsen er mindre enn 'md'</span>
        )}
        {size.gt.md && (
          <span>Vises når skjermstørrelsen er større enn 'md'</span>
        )}
      </div>
    );
  };

  it('matcher snapshot', () => {
    const wrapper = shallow(<TestHook />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('skal gi riktig størrelse basert på skjermbredde', () => {
    const wrapper = mount(<TestHook />);
    expect(wrapper.html()).toEqual(
      "<div><span>Vises når skjermstørrelsen er større enn 'md'</span></div>"
    );

    act(() => {
      global.innerWidth = 500;
      window.dispatchEvent(new Event('resize'));
    });
    expect(wrapper.html()).toEqual(
      "<div><span>Vises når skjermstørrelsen er 'md'</span></div>"
    );

    act(() => {
      global.innerWidth = 200;
      window.dispatchEvent(new Event('resize'));
    });
    expect(wrapper.html()).toEqual(
      "<div><span>Vises når skjermstørrelsen er mindre enn 'md'</span></div>"
    );
  });
});
