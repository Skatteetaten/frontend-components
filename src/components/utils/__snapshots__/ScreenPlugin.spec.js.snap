// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test av ScreenPlugin matcher snapshot 1`] = `
<div>
  <span>
    Vises når skjermstørrelsen er mindre enn 'md'
  </span>
</div>
`;
